### ISSUE TITLE
Fitnesse headless chrome, hangs often: Broken pipe.

### ISSUE THREAD

**Issue by rvanderboom** [fhoeben/hsac-fitnesse-fixtures#375]
Hello,

I use fitnesse (chrome headless, running in chrome browser on kubernetes, for some time, and like broad and customizability.

However from day one test suites sometimes hang (see error below).  I have tried so many things but no improvement at all.
Hope someone has some advice how to fix it, even if it is overwriting some classes.

Wat i tried an know:
- Run on java8, 11, 14, issue the same.
- Put on more cpu and memory (now 4G, but it is using only 500M according to kubernetes).
- Upgraded to latest version of all fitnesse related dependencies (older and new versions same issue).
- Used different ports (now 9090), no difference.
- Used different settings for seconds before timeout, no difference
- Added waits between steps, no difference.
- tried different settings for: seleniumDefaultTimeout, sun.net.client.defaultConnectTimeout, 
   sun.net.client.defaultReadTimeout
- Tried to find out if it is a certain instruction: no it looks like on different points it can hangup 

What i do know:
- It looks like more tests in a test suite, hangs up quicker. However sometimes even with 4 test in a suite it hangs.

```
Starting FitNesse on port: 9090
Test system terminated with exception [java.net.SocketException: Broken pipe (Write failed)]
SEVERE: Exception rendering #parse(testPage.vm) at skeleton.vm[line 55, column 4] [Broken pipe (Write failed)]
java.net.SocketException: Broken pipe (Write failed)
	at java.base/java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110)
	at java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:138)
	at fitnesse.FitNesseExpediter.send(FitNesseExpediter.java:63)
	at fitnesse.http.ChunkedResponse.add(ChunkedResponse.java:60)
	at fitnesse.http.ChunkedResponse.add(ChunkedResponse.java:45)
	at fitnesse.http.ChunkedResponse$1.write(ChunkedResponse.java:129)
	at org.apache.velocity.runtime.parser.node.ASTText.render(ASTText.java:115)
	at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:423)
	at org.apache.velocity.runtime.directive.Parse.render(Parse.java:279)
	at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:301)
	at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:423)
	at org.apache.velocity.Template.merge(Template.java:358)
	at org.apache.velocity.Template.merge(Template.java:262)
	at fitnesse.html.template.HtmlPage.render(HtmlPage.java:83)
	at fitnesse.responders.run.SuiteResponder.doSending(SuiteResponder.java:127)
	at fitnesse.responders.ChunkingResponder.startSending(ChunkingResponder.java:74)
	at fitnesse.http.ChunkedResponse.sendTo(ChunkedResponse.java:29)
	at fitnesse.FitNesseExpediter.sendResponse(FitNesseExpediter.java:86)
	at fitnesse.FitNesseExpediter.run(FitNesseExpediter.java:52)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by fhoeben** [comment]
Just to get the context right, you get this error when:
 1. running tests from your browser (i.e. clicking 'Test' or 'Suite' from the wiki), 
 2. the wiki server is running in Kubernetes (which OS on the node, could you share the container definition, e.g. Dockerfile)
 2. the browser you run from is on your local machine (or is that a headless chrome instance)
 3. the tests use this project to test a web application (e.g. BrowserTest to run tests using a headless chrome on kubernetes)
 4. is the headless chrome on the same node/container as the wiki server, or is that a separate container/node?

It might be related to, or even the same as:
 - https://github.com/unclebob/fitnesse/issues/1187
 - https://github.com/unclebob/fitnesse/issues/1283

Those are all issues for FitNesse in general, not this project specifically. Do you have any reason to assume the error has anything to do with this project's extensions of FitNesse?

Do you get this (or similar) issues when you run your tests directly with a docker container (i.e. without starting the wiki) using `hsac/fitnesse-fixtures-test-jre8-chrome` (which includes a (non-headless) chrome)? Along the lines with what is described at https://github.com/fhoeben/hsac-fitnesse-fixtures#running-using-docker.

**Comment by rvanderboom** [comment]
hey since it my free weekend i can answer all except deliver the dockerfile, will do next week:

1. running tests from your browser (i.e. clicking 'Test' or 'Suite' from the wiki)
    Correct running with Test or Suite from my local laptop chrome browser.
2. the wiki server is running in Kubernetes (which OS on the node, could you share the container definition, e.g. Dockerfile)
    Correct fitnesse standalone, wiki server running on kunernetes, 1 pod.  
3. the browser you run from is on your local machine (or is that a headless chrome instance)
    Correct
4. the tests use this project to test a web application (e.g. BrowserTest to run tests using a headless chrome on kubernetes)
     All correct, not sure but indeed seems like only browserTest has this issue, http/json api tests always work
5. is the headless chrome on the same node/container as the wiki server, or is that a separate container/node?
    Yes everything is on the same container, fitnesse server, chromedriver with all supplied options —headless, some disablings   
    like i everywhere see advised, changing these flags (except headless of course) do not solve the issue

Will look into the issues and test running within the pod, thanks

**Comment by fhoeben** [comment]
While I haven't experienced this myself I fear the kubernetes ingress, or the OS inside the container running the wiki server, might be terminating the request when no data is sent to the client for 'too long'. FitNesse sends data to the client for each test table once it is completed. Tests with BrowserTest can be slower, certainly when they have a single large table (scenario calls don't break up the table) and/or the site being tested is slow. This can mean there is quite a while between the HTML being sent to the client for the start of a test page, and the next data when the table is complete.

This would explain the error you see, and would be consistent with not having it for with JSON tests (that tend to be faster resulting in quicker feedback). 

You could see whether the time a connection can be open in your kubernetes environment without data being sent can be configured. And/or whether the behaviour changes when you split your tests into smaller tables.

**Comment by rvanderboom** [comment]
hey thanks a lot, your suggestion made me test if fitnesse suite runs ok without nginx.
By running it locally with forwarding to the same pod (so skipping nginx) makes the suite run fully !!
Now the solution is however not there yet, since i changed together with our nginx expert, all configurations values to higher/unlimitted (see below), but no change at all, it still hangs after few minutes (less then 3)

Wat i also notice if it hangs, it hangs more often after rerunning the test. If pod is restarted then somehow it free something and at least runs little longer.. Don't know if thats interesting information however...

-----
    nginx.ingress.kubernetes.io/client_max_body_size: '0'.     (unlimitted)
    nginx.ingress.kubernetes.io/force-ssl-redirect: 'true'
    nginx.ingress.kubernetes.io/proxy-body-size: '0'.  (unlimitted)
    nginx.ingress.kubernetes.io/proxy-buffering: 'off'.    (tried with off and on)
    nginx.ingress.kubernetes.io/proxy-connect-timeout: 600s
    nginx.ingress.kubernetes.io/proxy-read-timeout: 600s
    nginx.ingress.kubernetes.io/proxy-send-timeout: 600s

**Comment by fhoeben** [comment]
Could it be that the AWS load balancer also has an (idle) timeout?
At https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/config-idle-timeout.html (for the classic) I found:
> If no data has been sent or received by the time that the idle timeout period elapses, the load balancer closes the connection. 

And at https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#connection-idle-timeout I found:
> To ensure that lengthy operations such as file uploads have time to complete, send at least 1 byte of data before each idle timeout period elapses, and increase the length of the idle timeout period as needed.

**Comment by rvanderboom** [comment]
mm yes we have done several adjustments, idle time load balancer now an hour, also added ngiinx timeouts through configuration-snippet...

Is improved now only we still get sometimes error
Could not send/receive data with SUT

Looks like it happens if 1 test fails (browser timeout - functional issue), but then it doesn't continue to next tests because of this error.

**Comment by rvanderboom** [comment]
mm seems to be stable now after tuning some more, so is combination of settings..
Thanks for advice !!

Do also have different question but don't know if better to make new issue for it..
Try to run 2 suites in parallel (within kube with failsafe plugin)...
But it fails on a command "quit"... Strange that tests doesn't start on a quit - can not connect, it should continue if chromedriver is not started right ?   Well maybe better to make new issue for this..
I try to debug why this is....

**Comment by rvanderboom** [comment]
hey again, i now found somethings that looks like a bug...
appart from the timeout and loadbalancer settings, SUT message seems to be an issue when running the suite in a different order and using the feature "open $url  in new tab".. This feature seems to hangup fitnesse tests...
And i found out if i first open a new tab with my own website, then do 'previous tab' and open the url there, this order works.

I do not known if it is fitnesse or browser related but it looks like i am not the only one:
https://community.appian.com/discussions/f/general/14031/fitnesse-stops-after-a-new-tab-is-opened

So after playing with different order, and opening first own website in new tab, and external website in previous tab it goes ok..

**Comment by fhoeben** [comment]
@rvanderboom I'm sorry but I don't feel I have enough information to do anything with the issue of the new tab like this. 

 * Do you see hanging tests when running locally, or only via Kubernetes?
 * If you can reproduce locally: can you share a sample test (suite) that shows the problem?

The appian discussion is not something I'm familiar with. I don't even know whether they were using BrowserTest.

**Comment by rvanderboom** [comment]
the tab issues is there also on local..
If i find a general scenario in it i will lett you know...
But will be difficult it seems like it depends on the order and waiting time to go to the next/previous tab.
I now open browsertab, new tab , previous tab, and next tab again..  Instead of only tab, new tab . And also seperate the browsertest in several steps (start more often with BrowserTest (where left off...)..
Maybe a tab that is put outside the main view (when new / next tab is used), is somehow killed due to an unexpected error or timeout by the site that was opened or so... But problem is that appart from SUT error no visual error can be seen (except that it hangs...)
Lett you known if i know more...

**Comment by rvanderboom** [comment]
mmm sorry no answer for a long time the issue stil is there sometimes and it indeeds happens when one browser test (table), takes to long handling everything..
I am already for some time making smaller table browser tests..
 However also more levels seem to increase the delays. So have to switch to procedure calls in stead of include of other pages i think..

**Comment by fhoeben** [comment]
What exactly do you mean with the start and endtime of a browser test? Do you mean the duration of a single test page, or a script table of a table on a page?
I usually run my suites via the jUnit runner on a CI server. In that case the runtime per page is logged (and I see start/end per page in the CI servers output)

<img width="1116" alt="Screenshot 2023-02-01 at 12 39 35" src="https://user-images.githubusercontent.com/5467859/216032780-0d83a428-3c10-4e7f-b527-82a8632a7bdb.png">
<img width="1071" alt="Screenshot 2023-02-01 at 12 38 57" src="https://user-images.githubusercontent.com/5467859/216032788-d3337cba-23a1-4f40-9d44-c4e5597324b3.png">

**Comment by rvanderboom** [comment]
Wel best i would like execution time of single instruction in a browserTest and als per storyboard within a table..
So that i can try to find which parts take long time, or which construction (more levels ?) cause this long delays.

**Comment by fhoeben** [comment]
Something you could easily do I suppose is use 'storyboard' in your tests and then check the timestamps of the captured images.
If you would create your own subclass of BrowserTest you could override `takeScreenshot` to capture time between invocations and log/output that. If you would then use 'storyboard' you would get the data you need I expect.

**Comment by rvanderboom** [comment]
hey, thanks that is a good advice , will do that !
